// hexo-is/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var fs = require('fs');
var hexoLog = require('hexo-log');
var path = require('path');
var util = require('util');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var fs__namespace = /*#__PURE__*/_interopNamespaceDefault(fs);

var name = "hexo-is";
var pkg = {
	name: name};

function isCurrentHelper(postPath = '/', strict) {
  if (typeof this.path !== 'string') return false;
  const currentPath = this.path.replace(/^[^/].*/, '/$&');
  if (strict) {
    if (postPath.endsWith('/')) postPath += 'index.html';
    postPath = postPath.replace(/^[^/].*/, '/$&');
    return currentPath === postPath;
  }
  postPath = postPath.replace(/\/index\.html$/, '/');
  if (postPath === '/') return currentPath === '/index.html';
  postPath = postPath.replace(/^[^/].*/, '/$&');
  return currentPath.startsWith(postPath);
}
function isHomeHelper() {
  var _this$page;
  return Boolean((_this$page = this.page) === null || _this$page === void 0 ? void 0 : _this$page.__index);
}
function isPostHelper() {
  const page = this.page || {};
  const src = page.full_source || '';
  const layout = page.layout || '';
  if (layout.startsWith('post')) return true;
  if (layout.startsWith('page')) return false;
  if (src) return Boolean(page.__post) && src.includes('_posts');
  return Boolean(page.__post);
}
function isPageHelper() {
  const page = this.page || {};
  const src = page.full_source || '';
  const layout = page.layout || '';
  if (layout.startsWith('post')) return false;
  if (layout.startsWith('page')) return true;
  if (src) return Boolean(page.__post) && !src.includes('_posts');
  return Boolean(page.__page);
}
function isArchiveHelper() {
  var _this$page2;
  return Boolean((_this$page2 = this.page) === null || _this$page2 === void 0 ? void 0 : _this$page2.archive);
}
function isYearHelper(year) {
  const page = this.page;
  if (!(page !== null && page !== void 0 && page.archive)) return false;
  if (year) return page.year === year;
  return Boolean(page.year);
}
function isMonthHelper(year, month) {
  const page = this.page;
  if (!(page !== null && page !== void 0 && page.archive)) return false;
  if (year) {
    if (month) return page.year === year && page.month === month;
    return page.month === year;
  }
  return Boolean(page.year && page.month);
}
function isCategoryHelper(category) {
  const page = this.page;
  if (category) return (page === null || page === void 0 ? void 0 : page.category) === category;
  return Boolean(page === null || page === void 0 ? void 0 : page.category);
}
function isTagHelper(tag) {
  const page = this.page;
  if (tag) return (page === null || page === void 0 ? void 0 : page.tag) === tag;
  return Boolean(page === null || page === void 0 ? void 0 : page.tag);
}
function internalIs(hexo) {
  const obj = {
    current: false,
    home: false,
    post: false,
    page: false,
    archive: false,
    year: false,
    month: false,
    category: false,
    tag: false,
    message: 'try using second argument'
  };
  if (!hexo || typeof hexo.page === 'undefined') return obj;
  return {
    current: isCurrentHelper.call(hexo),
    home: isHomeHelper.call(hexo),
    post: isPostHelper.call(hexo),
    page: isPageHelper.call(hexo),
    archive: isArchiveHelper.call(hexo),
    year: isYearHelper.call(hexo),
    month: isMonthHelper.call(hexo),
    category: isCategoryHelper.call(hexo),
    tag: isTagHelper.call(hexo)
  };
}

const log = typeof hexo !== 'undefined' ? hexo.log : hexoLog.logger({
  debug: false,
  silent: false
});
/**
 * @example
 * // run inside plugin or theme event
 * import hexoIs from 'hexo-is';
 * const hexo = this;
 * console.log(hexoIs(hexo)); // object or string
 * @param hexo
 * @returns
 */
const hexoIs = function (hexo) {
  if (typeof hexo === 'undefined') return;
  if (typeof hexo['page'] != 'undefined') return internalIs(hexo);
  if (typeof hexo['type'] != 'undefined') {
    const ix = internalIs(hexo);
    if (typeof ix[hexo['type']] != 'undefined') ix[hexo['type']] = true;
    return ix;
  }
};
/**
 * Dump variable to file
 * @param toDump
 */
function hexoIsDump(toDump, name = '') {
  if (name.length > 0) name = '-' + name;
  const dump = util.inspect(toDump, {
    showHidden: true,
    depth: null
  });
  const loc = path.join('tmp/hexo-is/dump' + name + '.txt');
  if (!fs__namespace.existsSync(path.dirname(loc))) {
    fs__namespace.mkdirSync(path.dirname(loc), {
      recursive: true
    });
  }
  fs__namespace.writeFileSync(loc, dump);
  log.log(`${pkg.name}: dump saved to: ${path.resolve(loc)}`);
}

exports.default = hexoIs;
exports.hexoIs = hexoIs;
exports.hexoIsDump = hexoIsDump;
