// sbg-utility/dist/config/_config.mjs
import fs from 'fs-extra';
import url from 'node:url';
import path from 'upath';
import yaml__default from 'yaml';
import 'ansi-colors';
import '../utils/logger.mjs';
import 'debug';
import '../utils/decode-url.mjs';
import '../utils/deepMerge.mjs';
import '../utils/encode-url.mjs';
import 'path';
import 'bluebird';
import 'minimatch';
import '../utils/filemanager/getAppRootDir.mjs';
import '../utils/filemanager/path-utility.mjs';
import { writefile } from '../utils/filemanager/writefile.mjs';
import '../utils/generate-exports.mjs';
import '../utils/hash.mjs';
import '../utils/hash/getChecksum.mjs';
import '../utils/JSON.mjs';
import '../utils/JSON-serializer.mjs';
import '../utils/LocalStorageSharedPreferences.mjs';
import 'os';
import '../utils/scheduler.mjs';
import '../utils/nodeWorkspaceHelper.mjs';
import 'nunjucks';
import { orderKeys } from '../utils/object.mjs';
import '../utils/promisify.mjs';
import '../utils/_punycode.mjs';
import 'fs';
import 'stream';
import { getDefaultConfig } from './default-config.mjs';

const __filename = url.fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const configFileJSON = path.join(__dirname, '_config.json');
if (!fs.existsSync(configFileJSON)) fs.writeFileSync(configFileJSON, '{}');
let settledConfig = getDefaultConfig();
/**
 * find `_config.yml`
 * @param fileYML path to file `_config.yml` or working directory
 */
function fetchConfig(fileYML) {
  if (!fileYML) {
    fileYML = path.join(process.cwd(), '_config.yml');
  } else if (!fileYML.endsWith('_config.yml')) {
    fileYML += '/_config.yml';
  }
  const configYML = yaml__default.parse(fs.readFileSync(path.resolve(fileYML), 'utf-8'));
  setConfig(orderKeys(configYML));
  writefile(configFileJSON, JSON.stringify(configYML, null, 2));
}
// fetch _config.yml first init
// fetchConfig(join(process.cwd(), '_config.yml'));
/**
 * Config setter
 * * useful for jest
 * @param obj
 */
function setConfig(obj) {
  settledConfig = Object.assign(settledConfig || {}, obj);
  return getConfig();
}
/**
 * Config getter
 * * useful for jest
 * @returns
 */
function getConfig() {
  settledConfig.deploy = Object.assign(settledConfig.deploy || {}, deployConfig());
  return settledConfig;
}
/**
 * get deployment config
 * @returns
 */
function deployConfig() {
  var _settledConfig$deploy2;
  let deployDir = settledConfig.deploy_dir;
  if (!deployDir) {
    var _settledConfig$deploy;
    // fallback get from deploy.type
    const deployType = ((_settledConfig$deploy = settledConfig.deploy) === null || _settledConfig$deploy === void 0 ? void 0 : _settledConfig$deploy.type) || 'git';
    deployDir = path.join(settledConfig.cwd || process.cwd(), `.deploy_${deployType}`);
  }
  // subfolder - assign deploy.folder
  if ((_settledConfig$deploy2 = settledConfig.deploy) !== null && _settledConfig$deploy2 !== void 0 && _settledConfig$deploy2.folder) {
    deployDir = path.join(deployDir, settledConfig.deploy.folder);
  }
  // Also set deploy_dir on config for consistency
  settledConfig.deploy_dir = deployDir;
  return {
    deployDir
  };
}
/**
 * common ignore files
 * @example
 * const config = getConfig();
 * const excludes = Array.isArray(config.exclude) ? config.exclude : [];
 * excludes.push(...commonIgnore);
 */
const commonIgnore = ['**/yandex_*.html',
// skip yandex verification file
// '**/comments.html',
// '**/disqus-comments.html',
// '**/comment.html', // skip comment.html
'**/favicon.html',
// skip favicon
'**/404.html',
// skip 404
'**/node_modules/**',
// skip node_modules
'**/tmp/**',
// skip tmp
'**/build/**',
// skip build
'**/.cache/**',
// skip common cache folder
'**/.vscode/**',
// skip vscode configuration folder
'**/.frontmatter/**',
// skip frontmatter vscode extension
'**/pinterest-*.html',
// skip pinterest verification file
'**/_*.standalone.{js,ts}',
// skip _filename.standalone.js
'**/desktop.ini',
// windows desktop.ini
'**/node_modules/**',
// node_modules
'**/.frontmatter/**',
// vscode frontmatter plugin
'**/.git*/**',
// any git configs
'**/.*' // skip dots
];
/**
 * array of config.exclude, config.ignore
 */
const projectIgnores = [...(getConfig().skip_render || []), ...(getConfig().ignore || [])];

export { commonIgnore, deployConfig, fetchConfig, getConfig, projectIgnores, setConfig };
