// sbg-utility/dist/utils/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var indexExports = require('./index-exports.cjs');
var getChecksum = require('./hash/getChecksum.cjs');
var array = require('./array.cjs');
var chain = require('./chain.cjs');
var date = require('./date.cjs');
var debug = require('./debug.cjs');
var decodeUrl = require('./decode-url.cjs');
var deepMerge = require('./deepMerge.cjs');
var encodeUrl = require('./encode-url.cjs');
var environment = require('./environment.cjs');
var copy = require('./filemanager/copy.cjs');
var del = require('./filemanager/del.cjs');
var driveLetterUtils = require('./filemanager/driveLetterUtils.cjs');
var emptyDir = require('./filemanager/emptyDir.cjs');
var getAppRootDir = require('./filemanager/getAppRootDir.cjs');
var images = require('./filemanager/images.cjs');
var pathUtility = require('./filemanager/path-utility.cjs');
var readDir = require('./filemanager/readDir.cjs');
var readfile = require('./filemanager/readfile.cjs');
var stream = require('./filemanager/stream.cjs');
var writefile = require('./filemanager/writefile.cjs');
var index = require('./filemanager/index.cjs');
var generateExports = require('./generate-exports.cjs');
var hash = require('./hash.cjs');
var jest = require('./jest.cjs');
var JSON = require('./JSON.cjs');
var JSONSerializer = require('./JSON-serializer.cjs');
var JSONSerializerBrowser = require('./JSON-serializer-browser.cjs');
var LocalStorageSharedPreferences = require('./LocalStorageSharedPreferences.cjs');
var logger = require('./logger.cjs');
var MemorySharedPreferences = require('./MemorySharedPreferences.cjs');
var nodeWorkspaceHelper = require('./nodeWorkspaceHelper.cjs');
var noop = require('./noop.cjs');
var nunjucksEnv = require('./nunjucks-env.cjs');
var object = require('./object.cjs');
var persistentCache = require('./persistent-cache.cjs');
var promise = require('./promise.cjs');
var promisify = require('./promisify.cjs');
var _punycode = require('./_punycode.cjs');
var scheduler = require('./scheduler.cjs');
var semver = require('./semver.cjs');
var SharedPreferences = require('./SharedPreferences.cjs');
var stream$1 = require('./stream.cjs');
var string = require('./string.cjs');
var uuid = require('./uuid.cjs');



exports.default = indexExports;
exports.getChecksum = getChecksum;
exports.arrayOfObjUniq = array.arrayOfObjUniq;
exports.array_flatten = array.array_flatten;
exports.array_random = array.array_random;
exports.array_remove_empty = array.array_remove_empty;
exports.array_shuffle = array.array_shuffle;
exports.array_shuffle_swap = array.array_shuffle_swap;
exports.array_unique = array.array_unique;
exports.rand = array.rand;
exports.chain = chain.chain;
exports.toMilliseconds = date.toMilliseconds;
exports.debug = debug.debug;
exports.sbgDebug = debug.sbgDebug;
exports.decodeURL = decodeUrl.decodeURL;
exports.deepmerge = deepMerge.deepmerge;
exports.encodeURL = encodeUrl.encodeURL;
exports.getBinaryPath = environment.get_binary_path;
exports.get_binary_path = environment.get_binary_path;
exports.isClass = environment.isClass;
exports.isdev = environment.isdev;
exports.copyPath = copy.copyPath;
exports.del = del.del;
exports.fixDriveLetter = driveLetterUtils.fixDriveLetter;
exports.emptyDir = emptyDir.emptyDir;
exports.getAppRootDir = getAppRootDir.getAppRootDir;
exports.image_base64_encode = images.image_base64_encode;
exports.isWindows = pathUtility.isWindows;
exports.joinPath = pathUtility.normalizePath;
exports.joinSolve = pathUtility.joinSolve;
exports.normalizePath = pathUtility.normalizePath;
exports.normalizePathUnix = pathUtility.normalizePathUnix;
exports.pathJoin = pathUtility.normalizePath;
exports.removeCwd = pathUtility.removeCwd;
exports.trueCasePath = pathUtility.trueCasePath;
exports.trueCasePathSync = pathUtility.trueCasePathSync;
exports.readDir = readDir.readDir;
exports.readDirAsync = readDir.readDirAsync;
exports.readFile = readfile.readfile;
exports.readfile = readfile.readfile;
exports.createWriteStream = stream.createWriteStream;
exports.writeFile = writefile.writefile;
exports.writefile = writefile.writefile;
exports.isAsset = index.isAsset;
exports.isMarkdown = index.isMarkdown;
exports.generateExports = generateExports.generateExports;
exports.data_to_hash = hash.data_to_hash;
exports.data_to_hash_sync = hash.data_to_hash_sync;
exports.file_to_hash = hash.file_to_hash;
exports.folder_to_hash = hash.folder_to_hash;
exports.md5 = hash.md5;
exports.md5File = hash.md5File;
exports.md5FileSync = hash.md5FileSync;
exports.url_to_hash = hash.url_to_hash;
exports.areWeTestingWithJest = jest.areWeTestingWithJest;
exports.jsonParseWithCircularRefs = JSON.jsonParseWithCircularRefs;
exports.jsonStringifyWithCircularRefs = JSON.jsonStringifyWithCircularRefs;
exports.fromJSON = JSONSerializer.fromJSON;
exports.parse = JSONSerializer.parse;
exports.stringify = JSONSerializer.stringify;
exports.toJSON = JSONSerializer.toJSON;
exports.fromJSONBrowser = JSONSerializerBrowser.fromJSONBrowser;
exports.jsonParseWithCircularRefsBrowser = JSONSerializerBrowser.jsonParseWithCircularRefsBrowser;
exports.jsonStringifyWithCircularRefsBrowser = JSONSerializerBrowser.jsonStringifyWithCircularRefsBrowser;
exports.parseBrowser = JSONSerializerBrowser.parseBrowser;
exports.stringifyBrowser = JSONSerializerBrowser.stringifyBrowser;
exports.toJSONBrowser = JSONSerializerBrowser.toJSONBrowser;
exports.LocalStorageSharedPreferences = LocalStorageSharedPreferences.LocalStorageSharedPreferences;
exports.Logger = logger.Logger;
exports.disableConsoleLogForJest = logger.disableConsoleLogForJest;
exports.MemorySharedPreferences = MemorySharedPreferences.MemorySharedPreferences;
exports.cmd = nodeWorkspaceHelper.cmd;
exports.findYarnRootWorkspace = nodeWorkspaceHelper.findYarnRootWorkspace;
exports.resolveCommand = nodeWorkspaceHelper.resolveCommand;
exports.noop = noop.noop;
exports.envNunjucks = nunjucksEnv.envNunjucks;
exports.getObjectProperty = object.getObjectProperty;
exports.orderKeys = object.orderKeys;
exports.persistentCache = persistentCache.persistentCache;
exports.safeCb = persistentCache.safeCb;
exports.delay = promise.delay;
exports.promisify = promisify.promisify;
exports.punycode = _punycode.punycode;
exports.bindProcessExit = scheduler.bindProcessExit;
exports.scheduler = scheduler.scheduler;
exports.semver = semver.semver;
exports.semverIncrement = semver.semverIncrement;
exports.SharedPreferences = SharedPreferences.SharedPreferences;
exports.createDuplexStream = stream$1.createDuplexStream;
exports.bufferToString = string.bufferToString;
exports.capitalize = string.capitalize;
exports.capitalizer = string.capitalizer;
exports.cleanString = string.cleanString;
exports.escapeRegex = string.escapeRegex;
exports.isValidHttpUrl = string.isValidHttpUrl;
exports.replacePath = string.replacePath;
exports.slugify = string.slugify;
exports.streamToString = string.streamToString;
exports.makeid = uuid.makeid;
