// sbg-utility/dist/utils/LocalStorageSharedPreferences.mjs
import { LocalStorage } from 'node-localstorage';

if (typeof localStorage === 'undefined' || localStorage === null) {
  // Only assign polyfill in Node.js
  if (typeof process !== 'undefined' && process.versions && process.versions.node) {
    globalThis.localStorage = new LocalStorage('./scratch');
  }
}
/**
 * Implementation of SharedPreferences using localStorage.
 */
class LocalStorageSharedPreferences {
  namespace;
  /**
   * Creates a new LocalStorageSharedPreferences instance.
   * @param namespace Namespace for all keys, similar to Android's SharedPreferences name. Required.
   */
  constructor(namespace) {
    this.namespace = namespace;
    if (!namespace) {
      throw new Error('Namespace is required for LocalStorageSharedPreferences');
    }
  }
  /**
   * Retrieves a string value for the given key.
   * @param key The key to look up.
   * @param defaultValue The value to return if the key is not found.
   */
  getString(key, defaultValue) {
    const value = localStorage.getItem(this.namespace + ':' + key);
    if (value === null || value === undefined) {
      return defaultValue;
    }
    return value;
  }
  /**
   * Stores a string value for the given key.
   * @param key The key to store.
   * @param value The string value to store.
   */
  putString(key, value) {
    localStorage.setItem(this.namespace + ':' + key, value);
  }
  /**
   * Retrieves an integer value for the given key.
   * @param key The key to look up.
   * @param defaultValue The value to return if the key is not found.
   */
  getInt(key, defaultValue) {
    const value = localStorage.getItem(this.namespace + ':' + key);
    if (value === null || value === undefined) {
      return defaultValue;
    }
    const parsed = parseInt(value, 10);
    return isNaN(parsed) ? defaultValue : parsed;
  }
  /**
   * Stores an integer value for the given key.
   * @param key The key to store.
   * @param value The integer value to store.
   */
  putInt(key, value) {
    localStorage.setItem(this.namespace + ':' + key, value.toString());
  }
  /**
   * Retrieves a boolean value for the given key.
   * @param key The key to look up.
   * @param defaultValue The value to return if the key is not found.
   */
  getBoolean(key, defaultValue) {
    const value = localStorage.getItem(this.namespace + ':' + key);
    if (value === null || value === undefined) {
      return defaultValue;
    }
    if (value === 'true') return true;
    if (value === 'false') return false;
    return defaultValue;
  }
  /**
   * Stores a boolean value for the given key.
   * @param key The key to store.
   * @param value The boolean value to store.
   */
  putBoolean(key, value) {
    localStorage.setItem(this.namespace + ':' + key, value ? 'true' : 'false');
  }
  /**
   * Removes the value for the given key.
   * @param key The key to remove.
   */
  remove(key) {
    localStorage.removeItem(this.namespace + ':' + key);
  }
  /**
   * Clears all keys and values.
   */
  clear() {
    // Only remove keys with the namespace
    for (let i = localStorage.length - 1; i >= 0; i--) {
      const key = localStorage.key(i);
      if (key && key.startsWith(this.namespace + ':')) {
        localStorage.removeItem(key);
      }
    }
  }
}

export { LocalStorageSharedPreferences, LocalStorageSharedPreferences as default };
