// sbg-utility/dist/utils/logger.mjs
import fs from 'fs-extra';
import path__default from 'path';
import slugify from 'slugify';
import path from 'upath';
import { fileURLToPath } from 'url';
import { getConfig } from '../config/_config.mjs';
import '../config/config-wrapper.mjs';
import '../config/default-config.mjs';
import 'bluebird';
import 'minimatch';
import './filemanager/getAppRootDir.mjs';
import './filemanager/path-utility.mjs';
import { writefile } from './filemanager/writefile.mjs';
import { areWeTestingWithJest } from './jest.mjs';

const __filename = fileURLToPath(import.meta.url);
path__default.dirname(__filename);
let FOLDER = path.join(process.cwd(), 'tmp/logs');
let cwd = process.cwd();
/**
 * Disables console.log when running tests with Jest.
 * Logs messages to a file instead.
 */
function disableConsoleLogForJest() {
  if (areWeTestingWithJest()) {
    const log = console.log;
    console.log = function (...args) {
      if (typeof getConfig === 'function') {
        const cfg = getConfig();
        FOLDER = path.join(cfg.cwd, 'tmp/logs/');
        cwd = cfg.cwd;
      }
      const stack = (new Error('').stack || '').split(/\r?\n/gm);
      let msg = (stack || [])[3] || '';
      if (msg.includes(__filename)) {
        msg = (stack || [])[2] || '';
      }
      const filename = slugify(path.toUnix(msg).replace(path.toUnix(cwd), ''), {
        lower: true,
        trim: true,
        replacement: '-',
        strict: true
      });
      const header = `\n\n ${new Date()} \n\n`;
      const write = writefile(path.join(FOLDER, filename + '.log'), header + args.join('\n\n'), {
        append: true
      });
      log(write.file);
    };
  }
}
//const _log = typeof hexo === 'undefined' ? console : Object.assign({ log: console.log }, hexo.log);
const _log = console;
/**
 * @example
 * const console = Logger
 * Logger.log('hello world'); // should be written in <temp folder>/logs/[trace-name].log
 */
class Logger {
  static log(...args) {
    _log.log(...args);
    this.tracer(...args);
  }
  static info(...args) {
    _log.info.apply(null, args);
    this.tracer(...args);
  }
  static error(...args) {
    _log.error.apply(null, args);
    this.tracer(...args);
  }
  static tracer(...args) {
    var _error$stack;
    const error = new Error();
    const split = (_error$stack = error.stack) === null || _error$stack === void 0 ? void 0 : _error$stack.split(/\r?\n/gm).map(str => {
      var _split2$;
      const split2 = str.trim().split(' ');
      return {
        name: split2[1],
        path: (_split2$ = split2[2]) === null || _split2$ === void 0 ? void 0 : _split2$.replace(/\\+/gm, '/').replace(/^\(/, '').replace(/\)$/, '')
        //trace: error.stack
      };
    });
    if (split) {
      split.splice(0, 3);
      let logfile;
      let templ;
      // Ensure split[0] and split[1] exist before accessing their properties
      if (split.length > 1 && split[0] && split[1] && typeof split[0].path === 'undefined' && typeof split[1].name === 'string' && typeof split[1].path === 'string' && split[1].path.includes('anonymous')) {
        const id = split[1].name;
        const filePath = split[0].name;
        const base = path.basename(filePath.split(':')[0].length === 1 ? filePath.split(':')[0] + ':' + filePath.split(':')[1] : filePath.split(':')[0]);
        logfile = path.join(FOLDER, slugify(id, {
          trim: true
        }) + '-' + slugify(base, {
          trim: true
        }) + '.log');
        if (!fs.existsSync(logfile)) {
          writefile(logfile, '');
        }
        templ = `${'='.repeat(20)}\nfile: ${filePath}\ndate: ${new Date()}\n${'='.repeat(20)}\n\n`;
        args.forEach(o => {
          if (o === null) o = 'null';
          if (typeof o === 'object') {
            try {
              o = JSON.stringify(o, null, 2);
            } catch {
              //
            }
          }
          templ += String(o) + '\n\n';
        });
        fs.appendFileSync(logfile, templ);
      }
      // Logger.log(split);
    }
  }
}

export { Logger, Logger as default, disableConsoleLogForJest };
