// sbg-utility/dist/utils/SharedPreferences.mjs
import { LocalStorageSharedPreferences } from './LocalStorageSharedPreferences.mjs';
import { MemorySharedPreferences } from './MemorySharedPreferences.mjs';

class SharedPreferences {
  store;
  constructor(options) {
    const defaults = {
      namespace: 'default',
      type: 'localStorage'
    };
    const mergedOptions = {
      ...defaults,
      ...options
    };
    if (mergedOptions.type === 'localStorage') {
      this.store = new LocalStorageSharedPreferences(mergedOptions.namespace);
    } else if (mergedOptions.type === 'memory') {
      this.store = new MemorySharedPreferences(mergedOptions.namespace);
    } else {
      throw new Error('Unsupported SharedPreferences type: ' + mergedOptions.type);
    }
  }
  getString(key, defaultValue) {
    return this.store.getString(key, defaultValue);
  }
  putString(key, value) {
    this.store.putString(key, value);
  }
  getInt(key, defaultValue) {
    return this.store.getInt(key, defaultValue);
  }
  putInt(key, value) {
    this.store.putInt(key, value);
  }
  getBoolean(key, defaultValue) {
    return this.store.getBoolean(key, defaultValue);
  }
  putBoolean(key, value) {
    this.store.putBoolean(key, value);
  }
  remove(key) {
    this.store.remove(key);
  }
  clear() {
    this.store.clear();
  }
}

export { SharedPreferences, SharedPreferences as default };
