// sbg-utility/dist/index-exports.cjs
'use strict';

var index$1 = require('./config/index.cjs');
var globals = require('./globals.cjs');
var index$2 = require('./gulp-utils/index.cjs');
var index = require('./utils/index.cjs');
var sitemapCrawler = require('./sitemap-crawler/sitemap-crawler.cjs');
var getChecksum = require('./utils/hash/getChecksum.cjs');
var _config = require('./config/_config.cjs');
var configWrapper = require('./config/config-wrapper.cjs');
var defaultConfig = require('./config/default-config.cjs');
var gulpCache = require('./gulp-utils/gulp-cache.cjs');
var gulpDom = require('./gulp-utils/gulp-dom.cjs');
var gulp_debug = require('./gulp-utils/gulp.debug.cjs');
var array = require('./utils/array.cjs');
var chain = require('./utils/chain.cjs');
var date = require('./utils/date.cjs');
var debug = require('./utils/debug.cjs');
var decodeUrl = require('./utils/decode-url.cjs');
var deepMerge = require('./utils/deepMerge.cjs');
var encodeUrl = require('./utils/encode-url.cjs');
var environment = require('./utils/environment.cjs');
var copy = require('./utils/filemanager/copy.cjs');
var del = require('./utils/filemanager/del.cjs');
var driveLetterUtils = require('./utils/filemanager/driveLetterUtils.cjs');
var emptyDir = require('./utils/filemanager/emptyDir.cjs');
var getAppRootDir = require('./utils/filemanager/getAppRootDir.cjs');
var images = require('./utils/filemanager/images.cjs');
var pathUtility = require('./utils/filemanager/path-utility.cjs');
var readDir = require('./utils/filemanager/readDir.cjs');
var readfile = require('./utils/filemanager/readfile.cjs');
var stream = require('./utils/filemanager/stream.cjs');
var writefile = require('./utils/filemanager/writefile.cjs');
var index$3 = require('./utils/filemanager/index.cjs');
var generateExports = require('./utils/generate-exports.cjs');
var hash = require('./utils/hash.cjs');
var jest = require('./utils/jest.cjs');
var JSON = require('./utils/JSON.cjs');
var JSONSerializer = require('./utils/JSON-serializer.cjs');
var JSONSerializerBrowser = require('./utils/JSON-serializer-browser.cjs');
var LocalStorageSharedPreferences = require('./utils/LocalStorageSharedPreferences.cjs');
var logger = require('./utils/logger.cjs');
var MemorySharedPreferences = require('./utils/MemorySharedPreferences.cjs');
var nodeWorkspaceHelper = require('./utils/nodeWorkspaceHelper.cjs');
var noop = require('./utils/noop.cjs');
var nunjucksEnv = require('./utils/nunjucks-env.cjs');
var object = require('./utils/object.cjs');
var persistentCache = require('./utils/persistent-cache.cjs');
var promise = require('./utils/promise.cjs');
var promisify = require('./utils/promisify.cjs');
var _punycode = require('./utils/_punycode.cjs');
var scheduler = require('./utils/scheduler.cjs');
var semver = require('./utils/semver.cjs');
var SharedPreferences = require('./utils/SharedPreferences.cjs');
var stream$1 = require('./utils/stream.cjs');
var string = require('./utils/string.cjs');
var uuid = require('./utils/uuid.cjs');

//

exports.config = index$1;
exports.globals = globals;
exports.gutils = index$2;
exports.utils = index;
exports.SiteMapCrawler = sitemapCrawler.SiteMapCrawlerCore;
exports.SiteMapCrawlerCore = sitemapCrawler.SiteMapCrawlerCore;
exports.sitemapCrawler = sitemapCrawler.sitemapCrawler;
exports.sitemapCrawlerAsync = sitemapCrawler.sitemapCrawlerAsync;
exports.getChecksum = getChecksum;
exports.commonIgnore = _config.commonIgnore;
exports.deployConfig = _config.deployConfig;
exports.fetchConfig = _config.fetchConfig;
exports.getConfig = _config.getConfig;
exports.main = _config;
exports.projectIgnores = _config.projectIgnores;
exports.setConfig = _config.setConfig;
exports.createConfig = configWrapper.createConfig;
exports.wrapper = configWrapper;
exports.defaults = defaultConfig;
exports.getDefaultConfig = defaultConfig.getDefaultConfig;
exports.getDefaultConfigYaml = defaultConfig.getDefaultConfigYaml;
exports.getShaFile = gulpCache.getShaFile;
exports.gulpCached = gulpCache.gulpCached;
exports.customPath = gulpDom.customPath;
exports.gulpDom = gulpDom.gulpDom;
exports.gulpDomPath = gulpDom.gulpDomPath;
exports.gulpDebug = gulp_debug.gulpDebug;
exports.gulpLog = gulp_debug.gulpLog;
exports.arrayOfObjUniq = array.arrayOfObjUniq;
exports.array_flatten = array.array_flatten;
exports.array_random = array.array_random;
exports.array_remove_empty = array.array_remove_empty;
exports.array_shuffle = array.array_shuffle;
exports.array_shuffle_swap = array.array_shuffle_swap;
exports.array_unique = array.array_unique;
exports.rand = array.rand;
exports.chain = chain.chain;
exports.toMilliseconds = date.toMilliseconds;
exports.debug = debug.debug;
exports.sbgDebug = debug.sbgDebug;
exports.decodeURL = decodeUrl.decodeURL;
exports.deepmerge = deepMerge.deepmerge;
exports.encodeURL = encodeUrl.encodeURL;
exports.getBinaryPath = environment.get_binary_path;
exports.get_binary_path = environment.get_binary_path;
exports.isClass = environment.isClass;
exports.isdev = environment.isdev;
exports.copyPath = copy.copyPath;
exports.del = del.del;
exports.fixDriveLetter = driveLetterUtils.fixDriveLetter;
exports.emptyDir = emptyDir.emptyDir;
exports.getAppRootDir = getAppRootDir.getAppRootDir;
exports.image_base64_encode = images.image_base64_encode;
exports.isWindows = pathUtility.isWindows;
exports.joinPath = pathUtility.normalizePath;
exports.joinSolve = pathUtility.joinSolve;
exports.normalizePath = pathUtility.normalizePath;
exports.normalizePathUnix = pathUtility.normalizePathUnix;
exports.pathJoin = pathUtility.normalizePath;
exports.removeCwd = pathUtility.removeCwd;
exports.trueCasePath = pathUtility.trueCasePath;
exports.trueCasePathSync = pathUtility.trueCasePathSync;
exports.readDir = readDir.readDir;
exports.readDirAsync = readDir.readDirAsync;
exports.readFile = readfile.readfile;
exports.readfile = readfile.readfile;
exports.createWriteStream = stream.createWriteStream;
exports.writeFile = writefile.writefile;
exports.writefile = writefile.writefile;
exports.isAsset = index$3.isAsset;
exports.isMarkdown = index$3.isMarkdown;
exports.generateExports = generateExports.generateExports;
exports.data_to_hash = hash.data_to_hash;
exports.data_to_hash_sync = hash.data_to_hash_sync;
exports.file_to_hash = hash.file_to_hash;
exports.folder_to_hash = hash.folder_to_hash;
exports.md5 = hash.md5;
exports.md5File = hash.md5File;
exports.md5FileSync = hash.md5FileSync;
exports.url_to_hash = hash.url_to_hash;
exports.areWeTestingWithJest = jest.areWeTestingWithJest;
exports.jsonParseWithCircularRefs = JSON.jsonParseWithCircularRefs;
exports.jsonStringifyWithCircularRefs = JSON.jsonStringifyWithCircularRefs;
exports.fromJSON = JSONSerializer.fromJSON;
exports.parse = JSONSerializer.parse;
exports.stringify = JSONSerializer.stringify;
exports.toJSON = JSONSerializer.toJSON;
exports.fromJSONBrowser = JSONSerializerBrowser.fromJSONBrowser;
exports.jsonParseWithCircularRefsBrowser = JSONSerializerBrowser.jsonParseWithCircularRefsBrowser;
exports.jsonStringifyWithCircularRefsBrowser = JSONSerializerBrowser.jsonStringifyWithCircularRefsBrowser;
exports.parseBrowser = JSONSerializerBrowser.parseBrowser;
exports.stringifyBrowser = JSONSerializerBrowser.stringifyBrowser;
exports.toJSONBrowser = JSONSerializerBrowser.toJSONBrowser;
exports.LocalStorageSharedPreferences = LocalStorageSharedPreferences.LocalStorageSharedPreferences;
exports.Logger = logger.Logger;
exports.disableConsoleLogForJest = logger.disableConsoleLogForJest;
exports.MemorySharedPreferences = MemorySharedPreferences.MemorySharedPreferences;
exports.cmd = nodeWorkspaceHelper.cmd;
exports.findYarnRootWorkspace = nodeWorkspaceHelper.findYarnRootWorkspace;
exports.resolveCommand = nodeWorkspaceHelper.resolveCommand;
exports.noop = noop.noop;
exports.envNunjucks = nunjucksEnv.envNunjucks;
exports.getObjectProperty = object.getObjectProperty;
exports.orderKeys = object.orderKeys;
exports.persistentCache = persistentCache.persistentCache;
exports.safeCb = persistentCache.safeCb;
exports.delay = promise.delay;
exports.promisify = promisify.promisify;
exports.punycode = _punycode.punycode;
exports.bindProcessExit = scheduler.bindProcessExit;
exports.scheduler = scheduler.scheduler;
exports.semver = semver.semver;
exports.semverIncrement = semver.semverIncrement;
exports.SharedPreferences = SharedPreferences.SharedPreferences;
exports.createDuplexStream = stream$1.createDuplexStream;
exports.bufferToString = string.bufferToString;
exports.capitalize = string.capitalize;
exports.capitalizer = string.capitalizer;
exports.cleanString = string.cleanString;
exports.escapeRegex = string.escapeRegex;
exports.isValidHttpUrl = string.isValidHttpUrl;
exports.replacePath = string.replacePath;
exports.slugify = string.slugify;
exports.streamToString = string.streamToString;
exports.makeid = uuid.makeid;
