// sbg-utility/dist/index.mjs
import './_virtual/global-file.mjs';
import * as indexExports from './index-exports.mjs';
export { default as fs } from 'fs-extra';
export { default as path } from 'upath';
export { default as getChecksum } from './utils/hash/getChecksum.mjs';
import * as _config from './config/_config.mjs';
export { _config as main };
export { commonIgnore, deployConfig, fetchConfig, getConfig, projectIgnores, setConfig } from './config/_config.mjs';
import * as configWrapper from './config/config-wrapper.mjs';
export { configWrapper as wrapper };
export { createConfig } from './config/config-wrapper.mjs';
import * as defaultConfig from './config/default-config.mjs';
export { defaultConfig as defaults };
export { getDefaultConfig, getDefaultConfigYaml } from './config/default-config.mjs';
export { getShaFile, gulpCached } from './gulp-utils/gulp-cache.mjs';
export { customPath, gulpDom, gulpDomPath } from './gulp-utils/gulp-dom.mjs';
export { gulpDebug, gulpLog } from './gulp-utils/gulp.debug.mjs';
export { SiteMapCrawlerCore as SiteMapCrawler, SiteMapCrawlerCore, sitemapCrawler, sitemapCrawlerAsync } from './sitemap-crawler/sitemap-crawler.mjs';
export { arrayOfObjUniq, array_flatten, array_random, array_remove_empty, array_shuffle, array_shuffle_swap, array_unique, rand } from './utils/array.mjs';
export { chain } from './utils/chain.mjs';
export { toMilliseconds } from './utils/date.mjs';
export { debug, sbgDebug } from './utils/debug.mjs';
export { decodeURL } from './utils/decode-url.mjs';
export { deepmerge } from './utils/deepMerge.mjs';
export { encodeURL } from './utils/encode-url.mjs';
export { get_binary_path as getBinaryPath, get_binary_path, isClass, isdev } from './utils/environment.mjs';
export { copyPath } from './utils/filemanager/copy.mjs';
export { del } from './utils/filemanager/del.mjs';
export { fixDriveLetter } from './utils/filemanager/driveLetterUtils.mjs';
export { emptyDir } from './utils/filemanager/emptyDir.mjs';
export { getAppRootDir } from './utils/filemanager/getAppRootDir.mjs';
export { image_base64_encode } from './utils/filemanager/images.mjs';
export { isWindows, normalizePath as joinPath, joinSolve, normalizePath, normalizePathUnix, normalizePath as pathJoin, removeCwd, trueCasePath, trueCasePathSync } from './utils/filemanager/path-utility.mjs';
export { readDir, readDirAsync } from './utils/filemanager/readDir.mjs';
export { readfile as readFile, readfile } from './utils/filemanager/readfile.mjs';
export { createWriteStream } from './utils/filemanager/stream.mjs';
export { writefile as writeFile, writefile } from './utils/filemanager/writefile.mjs';
export { isAsset, isMarkdown } from './utils/filemanager/index.mjs';
export { generateExports } from './utils/generate-exports.mjs';
export { data_to_hash, data_to_hash_sync, file_to_hash, folder_to_hash, md5, md5File, md5FileSync, url_to_hash } from './utils/hash.mjs';
export { areWeTestingWithJest } from './utils/jest.mjs';
export { jsonParseWithCircularRefs, jsonStringifyWithCircularRefs } from './utils/JSON.mjs';
export { fromJSON, parse, stringify, toJSON } from './utils/JSON-serializer.mjs';
export { fromJSONBrowser, jsonParseWithCircularRefsBrowser, jsonStringifyWithCircularRefsBrowser, parseBrowser, stringifyBrowser, toJSONBrowser } from './utils/JSON-serializer-browser.mjs';
export { LocalStorageSharedPreferences } from './utils/LocalStorageSharedPreferences.mjs';
export { Logger, disableConsoleLogForJest } from './utils/logger.mjs';
export { MemorySharedPreferences } from './utils/MemorySharedPreferences.mjs';
export { cmd, findYarnRootWorkspace, resolveCommand } from './utils/nodeWorkspaceHelper.mjs';
export { noop } from './utils/noop.mjs';
export { envNunjucks } from './utils/nunjucks-env.mjs';
export { getObjectProperty, orderKeys } from './utils/object.mjs';
export { persistentCache, safeCb } from './utils/persistent-cache.mjs';
export { delay } from './utils/promise.mjs';
export { promisify } from './utils/promisify.mjs';
export { punycode } from './utils/_punycode.mjs';
export { bindProcessExit, scheduler } from './utils/scheduler.mjs';
export { semver, semverIncrement } from './utils/semver.mjs';
export { SharedPreferences } from './utils/SharedPreferences.mjs';
export { createDuplexStream } from './utils/stream.mjs';
export { bufferToString, capitalize, capitalizer, cleanString, escapeRegex, isValidHttpUrl, replacePath, slugify, streamToString } from './utils/string.mjs';
export { makeid } from './utils/uuid.mjs';
import * as index from './utils/index.mjs';
export { index as utils };
import * as index$1 from './config/index.mjs';
export { index$1 as config };
import * as index$2 from './gulp-utils/index.mjs';
export { index$2 as gutils };
import * as globals from './globals.mjs';
export { globals };

// Pollyfill globalThis in Node.js
//

export { indexExports as default };
