// sbg-utility/dist/gulp-utils/gulp-dom.mjs
import jsdom from 'jsdom';
import PluginError from 'plugin-error';
import through2 from 'through2';
import path$1 from 'upath';
import 'ansi-colors';
import { Logger } from '../utils/logger.mjs';
import 'debug';
import '../utils/decode-url.mjs';
import '../utils/deepMerge.mjs';
import '../utils/encode-url.mjs';
import 'fs-extra';
import 'path';
import 'bluebird';
import 'minimatch';
import '../utils/filemanager/getAppRootDir.mjs';
import { trueCasePathSync } from '../utils/filemanager/path-utility.mjs';
import '../utils/generate-exports.mjs';
import '../utils/hash.mjs';
import '../utils/hash/getChecksum.mjs';
import '../utils/JSON.mjs';
import '../utils/JSON-serializer.mjs';
import '../utils/LocalStorageSharedPreferences.mjs';
import 'os';
import '../utils/scheduler.mjs';
import '../utils/nodeWorkspaceHelper.mjs';
import 'nunjucks';
import '../utils/promisify.mjs';
import '../utils/_punycode.mjs';
import 'fs';
import 'stream';

const pluginName = 'gulp-dom';
const path = {
  join: (...str) => path$1.toUnix(trueCasePathSync(path$1.join(...str))),
  dirname: str => path$1.toUnix(trueCasePathSync(path$1.dirname(str))),
  toUnix: str => path$1.toUnix(trueCasePathSync(str))
};
const customPath = path;
const gulpDomPath = path;
/**
 * gulp-dom
 * @param mutator callback
 * @returns
 * @example
 * const gulp = require('gulp');
    gulp.task('html', function() {
        return gulp.src('./src/index.html')
            .pipe(gulpDom(function(){
                return this.querySelectorAll('body')[0].setAttribute('data-version', '1.0');
            }))
            .pipe(gulp.dest('./public/'));
    });
 */
function gulpDom(mutator) {
  const stream = through2.obj(function (file, _enc, callback) {
    if (file.isNull()) {
      return callback(null, file);
    }
    if (file.isStream()) {
      return stream.emit('error', new PluginError(pluginName, 'Streaming not supported'));
    }
    if (file.isBuffer()) {
      try {
        const dom = new jsdom.JSDOM(file.contents.toString('utf8'));
        const mutated = mutator.call(dom.window.document, file.path);
        file.contents = Buffer.from(typeof mutated === 'string' ? mutated : dom.serialize());
        callback(null, file);
        dom.window.close();
      } catch (e) {
        if (e instanceof Error) {
          Logger.log(e.message);
        }
        Logger.log(pluginName, 'drop file', file.path);
        // drop file
        callback();
      }
    }
  });
  return stream;
}

export { customPath, gulpDom as default, gulpDom, gulpDomPath };
