// sbg-utility/dist/utils/filemanager/index.cjs
'use strict';

require('fs-extra');
require('path');
require('bluebird');
require('minimatch');
require('upath');
var getAppRootDir = require('./getAppRootDir.cjs');
var pathUtility = require('./path-utility.cjs');
require('../logger.cjs');

// filemanager
/**
 * is non-markdown file
 * @param assetPath
 * @returns
 */
const isAsset = assetPath => /.(js|css|scss|njk|ejs|png|jpe?g|gif|svg|webp|json|html|txt)$/.test(String(assetPath));
/**
 * is markdown file
 * @param markdownPath
 * @returns
 */
const isMarkdown = markdownPath => /.(md)$/i.test(String(markdownPath));

exports.getAppRootDir = getAppRootDir.getAppRootDir;
exports.isWindows = pathUtility.isWindows;
exports.joinPath = pathUtility.normalizePath;
exports.joinSolve = pathUtility.joinSolve;
exports.normalizePath = pathUtility.normalizePath;
exports.normalizePathUnix = pathUtility.normalizePathUnix;
exports.pathJoin = pathUtility.normalizePath;
exports.removeCwd = pathUtility.removeCwd;
exports.trueCasePath = pathUtility.trueCasePath;
exports.trueCasePathSync = pathUtility.trueCasePathSync;
exports.isAsset = isAsset;
exports.isMarkdown = isMarkdown;
